File: code/java/OpenCogCognitive/v1/src/main/java/com/factset/sdk/OpenCogCognitive/AtomSpace.java
package com.factset.sdk.OpenCogCognitive;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

/**
 * AtomSpace - Core knowledge representation system
 * 
 * The AtomSpace is a hypergraph database that stores knowledge as atoms.
 */
public class AtomSpace {
    private final Map<String, Atom> atoms;
    private final Map<String, Set<String>> atomsByType;
    private final Map<String, Set<String>> atomsByName;
    private final Map<String, Set<String>> incomingSet;
    
    public AtomSpace() {
        this.atoms = new ConcurrentHashMap<>();
        this.atomsByType = new ConcurrentHashMap<>();
        this.atomsByName = new ConcurrentHashMap<>();
        this.incomingSet = new ConcurrentHashMap<>();
    }
    
    public Node addNode(Node node) {
        if (!atoms.containsKey(node.getId())) {
            atoms.put(node.getId(), node);
            
            atomsByType.computeIfAbsent(node.getType(), k -> ConcurrentHashMap.newKeySet())
                      .add(node.getId());
            
            atomsByName.computeIfAbsent(node.getName(), k -> ConcurrentHashMap.newKeySet())
                      .add(node.getId());
        }
        return node;
    }
    
    public Link addLink(Link link) {
        if (!atoms.containsKey(link.getId())) {
            atoms.put(link.getId(), link);
            
            atomsByType.computeIfAbsent(link.getType(), k -> ConcurrentHashMap.newKeySet())
                      .add(link.getId());
            
            // Update incoming sets
            for (Atom atom : link.getOutgoing()) {
                incomingSet.computeIfAbsent(atom.getId(), k -> ConcurrentHashMap.newKeySet())
                          .add(link.getId());
            }
        }
        return link;
    }
    
    public Optional<Atom> getAtom(String atomId) {
        return Optional.ofNullable(atoms.get(atomId));
    }
    
    public List<Atom> getAtomsByType(String atomType) {
        Set<String> atomIds = atomsByType.getOrDefault(atomType, Collections.emptySet());
        List<Atom> result = new ArrayList<>();
        for (String id : atomIds) {
            Atom atom = atoms.get(id);
            if (atom != null) {
                result.add(atom);
            }
        }
        return result;
    }
    
    public List<Atom> getAtomsByName(String name) {
        Set<String> atomIds = atomsByName.getOrDefault(name, Collections.emptySet());
        List<Atom> result = new ArrayList<>();
        for (String id : atomIds) {
            Atom atom = atoms.get(id);
            if (atom != null) {
                result.add(atom);
            }
        }
        return result;
    }
    
    public List<Link> getIncomingSet(Atom atom) {
        Set<String> linkIds = incomingSet.getOrDefault(atom.getId(), Collections.emptySet());
        List<Link> result = new ArrayList<>();
        for (String id : linkIds) {
            Atom a = atoms.get(id);
            if (a instanceof Link) {
                result.add((Link) a);
            }
        }
        return result;
    }
    
    public boolean removeAtom(Atom atom) {
        if (atoms.containsKey(atom.getId())) {
            atoms.remove(atom.getId());
            
            Set<String> typeSet = atomsByType.get(atom.getType());
            if (typeSet != null) {
                typeSet.remove(atom.getId());
            }
            
            Set<String> nameSet = atomsByName.get(atom.getName());
            if (nameSet != null) {
                nameSet.remove(atom.getId());
            }
            
            incomingSet.remove(atom.getId());
            
            return true;
        }
        return false;
    }
    
    public List<Atom> getAllAtoms() {
        return new ArrayList<>(atoms.values());
    }
    
    public int size() {
        return atoms.size();
    }
    
    public void clear() {
        atoms.clear();
        atomsByType.clear();
        atomsByName.clear();
        incomingSet.clear();
    }
    
    @Override
    public String toString() {
        return String.format("AtomSpace(atoms=%d)", size());
    }
}

File: code/dotnet/OpenCogCognitive/v1/src/FactSet.SDK.OpenCogCognitive/AtomSpace.cs
using System;
using System.Collections.Generic;
using System.Collections.Concurrent;
using System.Linq;

namespace FactSet.SDK.OpenCogCognitive
{
    /// <summary>
    /// AtomSpace - Core knowledge representation system
    /// 
    /// The AtomSpace is a hypergraph database that stores knowledge as atoms.
    /// </summary>
    public class AtomSpace
    {
        private readonly ConcurrentDictionary<string, Atom> atoms;
        private readonly ConcurrentDictionary<string, HashSet<string>> atomsByType;
        private readonly ConcurrentDictionary<string, HashSet<string>> atomsByName;
        private readonly ConcurrentDictionary<string, HashSet<string>> incomingSet;
        private readonly object lockObject = new object();

        public AtomSpace()
        {
            atoms = new ConcurrentDictionary<string, Atom>();
            atomsByType = new ConcurrentDictionary<string, HashSet<string>>();
            atomsByName = new ConcurrentDictionary<string, HashSet<string>>();
            incomingSet = new ConcurrentDictionary<string, HashSet<string>>();
        }

        public Node AddNode(Node node)
        {
            if (!atoms.ContainsKey(node.Id))
            {
                atoms[node.Id] = node;

                lock (lockObject)
                {
                    if (!atomsByType.ContainsKey(node.Type))
                        atomsByType[node.Type] = new HashSet<string>();
                    atomsByType[node.Type].Add(node.Id);

                    if (!atomsByName.ContainsKey(node.Name))
                        atomsByName[node.Name] = new HashSet<string>();
                    atomsByName[node.Name].Add(node.Id);
                }
            }
            return node;
        }

        public Link AddLink(Link link)
        {
            if (!atoms.ContainsKey(link.Id))
            {
                atoms[link.Id] = link;

                lock (lockObject)
                {
                    if (!atomsByType.ContainsKey(link.Type))
                        atomsByType[link.Type] = new HashSet<string>();
                    atomsByType[link.Type].Add(link.Id);

                    // Update incoming sets
                    foreach (var atom in link.Outgoing)
                    {
                        if (!incomingSet.ContainsKey(atom.Id))
                            incomingSet[atom.Id] = new HashSet<string>();
                        incomingSet[atom.Id].Add(link.Id);
                    }
                }
            }
            return link;
        }

        public Atom GetAtom(string atomId)
        {
            atoms.TryGetValue(atomId, out var atom);
            return atom;
        }

        public List<Atom> GetAtomsByType(string atomType)
        {
            if (!atomsByType.TryGetValue(atomType, out var atomIds))
                return new List<Atom>();

            return atomIds.Select(id => atoms[id]).ToList();
        }

        public List<Atom> GetAtomsByName(string name)
        {
            if (!atomsByName.TryGetValue(name, out var atomIds))
                return new List<Atom>();

            return atomIds.Select(id => atoms[id]).ToList();
        }

        public List<Link> GetIncomingSet(Atom atom)
        {
            if (!incomingSet.TryGetValue(atom.Id, out var linkIds))
                return new List<Link>();

            return linkIds.Select(id => atoms[id] as Link).Where(l => l != null).ToList();
        }

        public bool RemoveAtom(Atom atom)
        {
            if (atoms.TryRemove(atom.Id, out _))
            {
                lock (lockObject)
                {
                    if (atomsByType.TryGetValue(atom.Type, out var typeSet))
                        typeSet.Remove(atom.Id);

                    if (atomsByName.TryGetValue(atom.Name, out var nameSet))
                        nameSet.Remove(atom.Id);

                    incomingSet.TryRemove(atom.Id, out _);
                }
                return true;
            }
            return false;
        }

        public List<Atom> GetAllAtoms()
        {
            return atoms.Values.ToList();
        }

        public int Size => atoms.Count;

        public void Clear()
        {
            atoms.Clear();
            atomsByType.Clear();
            atomsByName.Clear();
            incomingSet.Clear();
        }

        public override string ToString()
        {
            return $"AtomSpace(atoms={Size})";
        }
    }
}
